[Alvin Jiang Min Jun] iP

alvinjiang1 opened the pull request:
#HelloPanda
> "When I fall down, I get back up" - Sun Tzu's clumsy brother ([source)](https://www.merriam-webster.com/dictionary/sike)

HelloPanda is my version of the Duke to-do list.
it is:

- easy to learn
- ~~easy to use~~ **IDIOT PROOF**

All you need to do is,

1. start it
2. type in your commands
3. Magic 🪄 
4. _Watch as your tasks are added_

Features:

- [X]  Manage tasks 
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's the ```main``` method:

```java
public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
```

Commit:
Add Gradle support

alvinjiang1 committed:
Level 1

Greet, Echo, Exit

alvinjiang1 committed:
Level-2

Add, List

alvinjiang1 committed:
Level-3

Mark as Done

alvinjiang1 committed:
A-TextUiTesting & Level-5

Automated Text UI Testing and Handle Errors

alvinjiang1 committed:
Level-6

Delete

alvinjiang1 committed:
Add a save file behaviour

To do list created by user is not saved to any file.

The creating of a text output allows the user to make use of the to-do list they have created using our program.

Let's make use of the File and FileWriter classes to create and write to a file that the user can access.

The file can be accessed under iP/data/duke.txt

alvinjiang1 committed:
Add Date Functionality

To do list takes any string, and implements it as the "deadline" for activities.

Accepting and displaying date of activities would improve the implementation of Duke.

Let's make use of Java's LocalDate and DateTimeFormatter classes to read and write dates in specific formats.

Using Java's in-built classes allow us to store the Date as objects that can be accessed and reformatted

alvinjiang1 committed:
Update Event.java

alvinjiang1 committed:
Merge branch 'branch-Level-7'

alvinjiang1 committed:
Add read file capability

alvinjiang1 committed:
Merge branch 'branch-Level-7'

alvinjiang1 committed:
Update Duke.java

alvinjiang1 committed:
Add more OOP in code

The code mainly relies on the Duke class to parse information, read and write to file, and display output.

More OOP will help to make the code more succint and organised.

Let's add a few additional classes Parser, Storage, TaskList, Ui which each modify a portion of the bot.

alvinjiang1 committed:
Organize classes into Duke package

Classes exist individually.

By grouping them under the same package, we can easily identify the relevant classes to the project.

alvinjiang1 committed:
Merge pull request #1 from alvinjiang1/add-gradle-support

Add Gradle support

alvinjiang1 committed:
Add Gradle and JUnit tests

Gradle is not supported and lack of JUnit testing.

Gradle was implemented to support JUnit testing.

JUnit testing is important to automate certain test cases, so that manual testing of functions do not need to be carried out.

alvinjiang1 committed:
Add the ability to find tasks

There is no way to look through the to-do list to find tasks.

Providing a way for users to look for existing tasks would be a useful upgrade.

Let's add a method as well as a case under the parser that supports the find command.

This increases the capability of our parser, and makes good use of existing classes.

alvinjiang1 committed:
Add javadocs

Added javadocs to the code

alvinjiang1 committed:
Coding Standard

Changed some code to follow the java coding standard

alvinjiang1 committed:
Merge branch 'find'

alvinjiang1 committed:
Merge branch 'coding-standard'

guokweijie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+public class Parser {
+
+    private TaskList tasks;
+    final static ArrayList<String> validCommands = new ArrayList<>(List.of("list", "mark", "unmark", "todo", "deadline", "event", "delete", "bye", "find"));
```
can add newline to shorten this line of code

guokweijie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+     */
+    public String parseCommand(String str) throws DukeException {
+        String command = str.split(" ", 2)[0];
+        if (validCommands.contains(command)) return command;
```
use egyptian braces

guokweijie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+        Ui ui = new Ui();
+        try {
+            switch (command) {
+                case "list":
```
case and switch should have the same level of indentation and be aligned

guokweijie commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,30 @@
+        TaskList tasks = new TaskList(new ArrayList<>(List.of("[T][ ] wake up",
+                "[D][ ] homework (by: Aug 24 2022)",
+                "[E][ ] lab (at: Jul 30 2023)")));
+        String desiredOutput = "[T][X] wake up\n" +
```
newline before the '+' operator

guokweijie reviewed (commented):
overall quite ok, just some small coding standard violations! keep up the good work!

marcusczh commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,73 @@
+     * @param taskList The taskList before a Task is added.
+     * @return ArrayList of type Task The taskList after the Task is added.
+     */
+    public ArrayList<Task> printAndStoreTask(ArrayList<Task> taskList) {
```
Perhaps instead of a function with a null return value, you can consider making it an abstract method

marcusczh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+     *                 an ArrayList of type Task, and stored as a field.
+     */
+    public TaskList(ArrayList<String> taskList) {
+        if (taskList.isEmpty()) tasks = new ArrayList<>();
```
You can consider using Egyptian-style braces here

marcusczh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    public TaskList(ArrayList<String> taskList) {
+        if (taskList.isEmpty()) tasks = new ArrayList<>();
+        ArrayList<Task> result = new ArrayList<>();
```
Perhaps you can consider using an explicit type argument for the ArrayList constructor

marcusczh commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,23 @@
+        TaskList t = new TaskList(new ArrayList<>(List.of("[T][ ] wake up",
+                "[D][ ] homework (by: Aug 24 2022)",
+                "[E][ ] lab (at: Jul 30 2023)")));
+        String desiredOutput = "1. [T][ ] wake up\n" +
+                "2. [D][ ] homework (by: Aug 24 2022)\n" +
+                "3. [E][ ] lab (at: Jul 30 2023)\n";
```
You may want to consider inserting line breaks before an operator to conform to the coding standard

marcusczh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @param i The index of the Task to be marked.
+     */
+    public void markTask(int i) {
```
Perhaps a more descriptive variable name like taskIndex may be suitable here

marcusczh reviewed (commented):
Other than a few minor issues, the overall code looks good and ready to merge! I really like how you used OOP to abstract out functionalities into classes!

Isaaclhy00 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
```
Perhaps 'by' could be more descriptive of its contents such as 'deadline' or 'deadlineDescription'

Isaaclhy00 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return String The String format of the DukeException Object.
+     */
+    @Override
```
is "Duke Exception detected necessary if the Duke Exception already displays an error message?

Isaaclhy00 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+     *
+     * @param filePath The file path to which the storage can read and write files.
+     */
+    public Storage(String filePath) {
```
Is this file path OS-independent as required? Could try using Paths.get() to fetch the relative file path then appending the input filePath to form the full file path for storage.

Isaaclhy00 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,73 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    /**
```
Seems like what you are going for here is an abstract method

Isaaclhy00 reviewed (commented):
Overall looks good and runs as expected. Week 3's project outline provides an Address Book example which exemplifies good OOP practices and organises all its classes and packages in a clear and concise manner. The Parser object should strictly handle parsing of the input and return a Command object. The Command object should take in (Tasklist, UI, Storage) and handle execution of the given command (handled by the TaskList object), store the updated task list (handled by the Storage object) and display the messages (handled by the UI object). Good to check out the Address Book example and align your code with the example given in the project outline :)

alvinjiang1 committed:
Add checkStyle

alvinjiang1 committed:
Add GUI to Duke

alvinjiang1 committed:
Merge branch 'branch-A-CheckStyle'

alvinjiang1 committed:
Add assertions into existing code

There are no assertions to test for correctness of code.

Assertions can help to detect programming errors before the code increases in length.

Let's add some Java assert statements to important portions of the code.

alvinjiang1 committed:
Improve Code Quality

Several methods are too long or do not follow SLAP principle.

By improving the code quality, the readability is greatly increased.

Let's write code that adheres to the SLAP principle and also abstract out components of long methods to different methods.

alvinjiang1 committed:
Add Anomoly Detection

When a new Event is added with the exact same date as a previous Event, it will be added to the To-Do List.

It would be helpful to add an ability to check for clashes in the Event dates, so that one do not commit to multiple events on the same day.

Let's add an additional method to check for this behaviour under each Event, and throw an exception when the same date is detected.

This can allow us to detect clashes while maintaining not adding the additional Event to our list.

alvinjiang1 committed:
Merge pull request #2 from alvinjiang1/branch-A-Assertions

Add assertions into existing code

alvinjiang1 committed:
Merge pull request #3 from alvinjiang1/branch-BCD-Extension

Add Anomoly Detection

alvinjiang1 committed:
Merge pull request #4 from alvinjiang1/branch-A-CodeQuality

Improve Code Quality

alvinjiang1 committed:
Edit readme and upload image

alvinjiang1 committed:
Improve Javadocs

alvinjiang1 committed:
Update README.md

alvinjiang1 committed:
Update TaskList.java
